### PR TITLE
Fix skipping unused shards with analyzer

### PR DESCRIPTION
**Pull request by vitlibar**
### Changelog category:
- Not for changelog

### Changelog entry:
Fix skipping unused shards with analyzer after https://github.com/ClickHouse/ClickHouse/pull/59606

This PR is also fixing tests broken in https://github.com/ClickHouse/ClickHouse/pull/59606:
`01527_dist_sharding_key_dictGet_reload`, `01940_custom_tld_sharding_key`, and [others](https://s3.amazonaws.com/clickhouse-test-reports/59606/5962ed0bd3ee89ba283c541982f60d13b1f010fa/stateless_tests__release__analyzer_.html)

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 12a71375b8a482ef9f64bcc970b5e1ae03758daa with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59741/12a71375b8a482ef9f64bcc970b5e1ae03758daa/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/59741/12a71375b8a482ef9f64bcc970b5e1ae03758daa/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Interpreters/ActionsDAG.cpp [hunk 1]
```diff
@@ -628,13 +628,17 @@ static ColumnWithTypeAndName executeActionForHeader(const ActionsDAG::Node * nod
             if (!array)
                 throw Exception(ErrorCodes::TYPE_MISMATCH,
                                 "ARRAY JOIN of not array nor map: {}", node->result_name);
-            res_column.column = array->getDataPtr()->cloneEmpty();
+            res_column.column = array->getDataPtr();
+            if (input_rows_count < array->size())
+                res_column.column = res_column.column->cloneResized(array->getOffsets()[input_rows_count - 1]);
```
Comment (novikd) [inline src/Interpreters/ActionsDAG.cpp:633]: This line is hard to understand. Why do we resite to the value of the offset?

Comment (vitlibar) [reply src/Interpreters/ActionsDAG.cpp:633]: This is how `arrayJoin()` works - it takes a `ColumnArray` and returns its data. So if only `input_rows_count` are to be used, then we should take only the first part of the data.

So this line is the same as `array->cloneResized(input_rows_count)->getDataPtr())` but in more optimized way. I'll add a comment about that here.

File: src/Interpreters/ActionsDAG.cpp [hunk 2]
```diff
@@ -768,7 +775,7 @@ ColumnsWithTypeAndName ActionsDAG::evaluatePartialResult(
                                         node->result_name);
 
                     if (node->type != ActionsDAG::ActionType::INPUT && has_all_arguments)
-                        node_to_column[node] = executeActionForHeader(node, std::move(arguments));
```
Comment (vitlibar) [inline src/Interpreters/ActionsDAG.cpp:771]: The problem was here. PR https://github.com/ClickHouse/ClickHouse/pull/59606 changed the [StorageDistributed::createSelector()](https://github.com/ClickHouse/ClickHouse/pull/59606/files#diff-b85e57e203f463cec4150a2721ba9bd58ad71dfae0698e57d3264ad7c5311934) function so now it returns an empty selector if the passed column is empty.

Function [StorageDistributed::skipUnusedShardsWithAnalyzer()](https://github.com/ClickHouse/ClickHouse/blob/b8f74b5b051eae34e3a68831add030b0a9978fbd/src/Storages/StorageDistributed.cpp#L1561) calls [evaluateExpressionOverConstantCondition()](https://github.com/ClickHouse/ClickHouse/blob/b8f74b5b051eae34e3a68831add030b0a9978fbd/src/Storages/StorageDistributed.cpp#L1607) to evaluate constant values of the sharding keys.
And then [evaluateExpressionOverConstantCondition()](https://github.com/ClickHouse/ClickHouse/blob/b8f74b5b051eae34e3a68831add030b0a9978fbd/src/Interpreters/evaluateConstantExpression.cpp#L673) calls [ActionsDAG::evaluatePartialResult()](https://github.com/ClickHouse/ClickHouse/blob/b8f74b5b051eae34e3a68831add030b0a9978fbd/src/Interpreters/evaluateConstantExpression.cpp#L664). However the [executeActionForHeader()](https://github.com/ClickHouse/ClickHouse/blob/b8f74b5b051eae34e3a68831add030b0a9978fbd/src/Interpreters/ActionsDAG.cpp#L608) function (which is called here) turned every column into empty which is ok for headers but not quite good for constant evaluation.

[Other overload of evaluateExpressionOverConstantCondition()](https://github.com/ClickHouse/ClickHouse/blob/b8f74b5b051eae34e3a68831add030b0a9978fbd/src/Interpreters/evaluateConstantExpression.h#L57)  always returns one row so I thought it's good to be consistent and make [this analyzer-specific overload of evaluateExpressionOverConstantCondition()](https://github.com/ClickHouse/ClickHouse/blob/b8f74b5b051eae34e3a68831add030b0a9978fbd/src/Interpreters/evaluateConstantExpression.h#L62) return one row too.

File: src/Interpreters/ActionsDAG.cpp (hunk 1 above)
Comment (novikd) [reply src/Interpreters/ActionsDAG.cpp:633]: Is it possible `input_rows_count` is zero here?

Comment (vitlibar) [reply src/Interpreters/ActionsDAG.cpp]: I've moved this very `ColumnArray`-specific code to the `ColumnArray` class itself, now it looks better.

**Review by novikd — APPROVED** [review]

**Pull request merged**